afew/filters/HeaderMatchingFilter.py

```diff
@@ -22,7 +22,14 @@ def __init__(self, database, **kwargs):
     def handle_message(self, message):
         if self.header is not None and self.pattern is not None:
             if not self._tag_blacklist.intersection(message.get_tags()):
-                value = message.get_header(self.header)
+                if not isinstance(self.header, list):
+                    self.header = [self.header]
+
+                try:
+                    value = next(filter(None, map(message.get_header, self.header)))
```

Comment:
This is pretty hard to read. I assume it'll call message.get_header() with every `header` configured in the filter, and filter out results that are an empty string (which can be the case if one of the headers doesn't exist)?

What's the next() in front of everything doing here? Why is `value` still a single element? What happens if the filter returns multiple elements?

Reply:
Yes, you read it right.

`value` is a single element, the first matching header. This is because the `HeaderMatchingFilter` only matches a string, it only has a pattern to match a string. Up until now the filter only matches to one header. For spam, as there are many spam flagging tools, you get different headers depending on the tool, thus you need to be able to search for many headers.
I assume, sysadmins don't run their many spam filters at one, they just pick a tool. Thus I pick the first not empty match(that is what `next` does), which might be the only match. Later I just let the the same logic of the filter run its way and match the pattern to the string. There I use the benefits of the regex. Spamassasing and Rspamd Mark with YES, their spam, bogofilter uses Spam.

Reply:
While it makes sense in the scope of `SpamFilter`, this changes semantics in `HeaderMatchingFilter` in a non-intuitive and undocumented way.

Can we expose this in `HeaderMatchingFilter` in a less confusing way, and have it documented?